sweep/store.go

```diff
@@ -0,0 +1,234 @@
+package sweep
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+)
+
+var (
+	// lastTxBucketKey is the key that points to a bucket containing a
+	// single item storing the last published tx.
+	//
+	// maps: lastTxKey -> serialized_tx
+	lastTxBucketKey = []byte("sweeper-last-tx")
+
+	// lastTxKey is the fixed key under which the serialized tx is stored.
+	lastTxKey = []byte("last-tx")
+
+	// txHashesBucketKey is the key that points to a bucket containing the
+	// hashes of all sweep txes that were published successfully.
+	//
+	// maps: txHash -> empty slice
+	txHashesBucketKey = []byte("sweeper-tx-hashes")
+
+	// utxnChainPrefix is the bucket prefix for nursery buckets.
+	utxnChainPrefix = []byte("utxn")
+
+	// utxnHeightIndexKey is the sub bucket where the nursery stores the
+	// height index.
+	utxnHeightIndexKey = []byte("height-index")
+
+	// utxnFinalizedKndrTxnKey is a static key that can be used to locate
+	// the nursery finalized kindergarten sweep txn.
+	utxnFinalizedKndrTxnKey = []byte("finalized-kndr-txn")
+)
+
+// SweeperStore stores published txes.
+type SweeperStore interface {
+	// IsOurTx determines whether a tx is published by us, based on its
+	// hash.
+	IsOurTx(hash chainhash.Hash) (bool, error)
+
+	// NotifyPublishTx signals that we are about to publish a tx.
+	NotifyPublishTx(*wire.MsgTx) error
+
+	// GetLastPublishedTx returns the last tx that we called NotifyPublishTx
+	// for.
+	GetLastPublishedTx() (*wire.MsgTx, error)
+}
+
+type sweeperStore struct {
+	db *channeldb.DB
+}
+
+// NewSweeperStore returns a new store instance.
+func NewSweeperStore(db *channeldb.DB) (SweeperStore, error) {
+	err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(
+			lastTxBucketKey,
+		)
+		if err != nil {
+			return err
+		}
+
+		if tx.Bucket(txHashesBucketKey) != nil {
+			return nil
+		}
+
+		txHashesBucket, err := tx.CreateBucket(txHashesBucketKey)
+		if err != nil {
+			return err
+		}
+
+		// Use non-existence of tx hashes bucket as a signal to migrate
+		// nursery finalized txes.
+		migrateTxHashes(tx, txHashesBucket)
+
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &sweeperStore{
+		db: db,
+	}, nil
+}
+
+// migrateTxHashes migrates nursery finalized txes to the tx hashes bucket. This
+// is not implemented as a database migration, to keep the downgrade path open.
+func migrateTxHashes(tx *bbolt.Tx, txHashesBucket *bbolt.Bucket) error {
+
+	// Iterate over all nursery chain buckets.
+	return tx.ForEach(func(name []byte, chainBucket *bbolt.Bucket) error {
```

Comment:
From my reading, this'll actually iterate over the _entire database_ rather than just the keys of the nursery. See my comment above re making a more targeted implementation via a new nursery method an an interface.

Reply:
It is iterating over all root buckets in the database scanning for nursery buckets. I don't think it is a performance problem. Wrt new nursery method, see comment above.